ksqldb-engine/src/main/java/io/confluent/ksql/util/SharedKafkaStreamsRuntimeImpl.java

```diff
@@ -0,0 +1,225 @@
+    kafkaStreams.start();
+  }
+
+  public void markSources(final QueryId queryId, final Set<SourceName> sourceNames) {
+    sources.put(queryId, sourceNames);
+  }
+
+  public void addQuery(
+          final QueryErrorClassifier errorClassifier,
+          final Map<String, Object> streamsProperties,
+          final PersistentQueriesInSharedRuntimesImpl persistentQueriesInSharedRuntimesImpl,
+          final QueryId queryId) {
+    this.errorClassifier = errorClassifier;
+    if (metadata.containsKey(queryId.toString())) {
+      kafkaStreams.removeNamedTopology(queryId.toString());
+    }
+    metadata.put(queryId.toString(), persistentQueriesInSharedRuntimesImpl);
+    LOG.debug("mapping {}", metadata);
+  }
+
+  public StreamsUncaughtExceptionHandler.StreamThreadExceptionResponse uncaughtHandler(
+      final Throwable e
+  ) {
+    QueryError.Type errorType = QueryError.Type.UNKNOWN;
+    try {
+      errorType = errorClassifier.classify(e);
+    } catch (final Exception classificationException) {
+      LOG.error("Error classifying unhandled exception", classificationException);
+    } finally {
+      // If error classification throws then we consider the error to be an UNKNOWN error.
+      // We notify listeners and add the error to the errors queue in the finally block to ensure
+      // all listeners and consumers of the error queue (e.g. the API) can see the error. Similarly,
+      // log in finally block to make sure that if there's ever an error in the classification
+      // we still get this in our logs.
+      final QueryError queryError =
+          new QueryError(
+              System.currentTimeMillis(),
+              Throwables.getStackTraceAsString(e),
+              errorType
+          );
+      for (PersistentQueriesInSharedRuntimesImpl query: metadata.values()) {
+        query.getListener().onError(metadata.get(query.getQueryId().toString()), queryError);
+      }
+      queryErrors.add(queryError);
+      LOG.error(
+          "Unhandled exception caught in streams thread {}. ({})",
+          Thread.currentThread().getName(),
+          errorType,
+          e
+      );
+    }
+    return StreamsUncaughtExceptionHandler.StreamThreadExceptionResponse.REPLACE_THREAD;
+  }
+
+  public KafkaStreams getKafkaStreams() {
+    return kafkaStreams;
+  }
+
+  public KafkaStreams.State state() {
+    return kafkaStreams.state();
+  }
+
+  public Collection<StreamsMetadata> allMetadata() {
+    return kafkaStreams.metadataForAllStreamsClients();
+  }
+
+  public Set<StreamsTaskMetadata> getTaskMetadata() {
+    return kafkaStreams.metadataForLocalThreads()
+                       .stream()
+                       .flatMap(t -> t.activeTasks().stream())
+                       .map(StreamsTaskMetadata::fromStreamsTaskMetadata)
+                       .collect(Collectors.toSet());
+  }
+
+  public void restart(final QueryId queryId) {
+    final KafkaStreamsNamedTopologyWrapper newKafkaStreams = kafkaStreamsBuilder
```

Comment:
Maybe we have discussed about this but I've forgot about our conclusions by then.. why we'd need to recreate a KS runtime and re-add the rest of the queries just to restart a new query? Could we just remove that query from the runtime and re-add it back, similarly to a `close(query)` followed by a `start(query)`?

Reply:
I think our discussion concluded with something like restart is not used, But it would be good to have a way to get a fresh runtime if necessary.

Reply:
If `restart` is used in the future, I think it is not necessary to get a fresh runtime still?

Reply:
We can just remove it if it's not used if you would prefer

Reply:
Ah, that works too :)